File: bin/Data/AUI/Periodic/LuaScripts/code_abuttongrid.lua
-- AButtonGrid application source code


function setup_uibuttongrid( layout)
    dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do 
        SubscribeToEvent(awid, "WidgetEvent", "HandleUibuttongridEvent")
    end

    local mybuttongrid = AButtonGrid:new(6,4)  -- make a scene:::view
    mybuttongrid:SetId( "AButtonGridDemo") -- tag it, in case we want to get it again later
    
    -- fill the buttongrid, feel the joy
    local row = 0
    while ( row < mybuttongrid:GetNumRows() ) do
        local col = 0
        while( col<mybuttongrid:GetNumColumns() ) do
            mybuttongrid:SetGridText (row, col, mybuttongrid:GetGridId(row, col));
            local bx = mybuttongrid:GetGridWidget( row, col );
            SubscribeToEvent(bx, "WidgetEvent", "HandleUibuttongridEvent" )
            col= col + 1
        end
        row = row + 1
    end
    
    local lpx = ALayoutParams:new()
    lpx:SetWidth (320)
    lpx:SetHeight(320)
    lpx:SetMinWidth(320)
    lpx:SetMinHeight(320)
    lpx:SetMaxWidth (320)
    lpx:SetMaxHeight(320)
    mybuttongrid:SetLayoutParams(lpx)

    local lower = layout:GetWidget("uibuttongridlower")
    local mysvc = layout:GetWidget("buttongridcontainer")
    mysvc:AddChildBefore(mybuttongrid, lower)
end

function HandleUibuttongridEvent( eventType, eventData)
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
        return
    end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uibuttongridcode" then
            AppLog( "AButtonGrid support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_abuttongrid.lua", widget:GetParent())
        elseif widget:GetId() ==  "uibuttongridlayout" then
            ViewCode ( "Scenes/layout_abuttongrid.ui.txt", widget:GetParent())
        else
            AppLog( "AButtonGrid support : " .. widget:GetId() .. " was pressed ")
        end
    end
    if widget:GetId() == "AButtonGridDemo" then
       AppLog( "AButtonGrid event : " .. widget:GetId() .. " got event= ".. eventData["Type"]:GetInt())
   end
end

File: Source/Samples/95_Periodic/code_abuttongrid.cpp
// AButtonGrid application source code
// some code from UISceneview2D sample program
#include <Urho3D/AUI/AButtonGrid.h>
#include <Urho3D/AUI/ALayout.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_abuttongrid( AWidget *layout)
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAbuttongridEvent ));

    AButtonGrid * mybuttongrid = new AButtonGrid(context_, 6, 4);  // make a mess
    mybuttongrid->SetId( "AButtonGridDemo"); // tag it, in case we want to get it again later
 
    // fill the buttongrid, feel the joy
    for ( unsigned row = 0; row<mybuttongrid->GetNumRows(); row++)
        for ( unsigned col = 0; col<mybuttongrid->GetNumColumns(); col++)
        {
            mybuttongrid->SetGridText (row, col, mybuttongrid->GetGridId(row, col));
            AWidget* bx = mybuttongrid->GetGridWidget( row, col );
            SubscribeToEvent(bx, E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAbuttongridEvent ));
        }

    ALayoutParams *lpx = new ALayoutParams(context_);
    lpx->SetWidth (320);
    lpx->SetHeight(320);
    lpx->SetMinWidth(320);
    lpx->SetMinHeight(320);
    lpx->SetMaxWidth (320);
    lpx->SetMaxHeight(320);
    mybuttongrid->SetLayoutParams(lpx);

    AWidget* lower = layout->GetWidget("uibuttongridlower");
    AWidget* mysvc = layout->GetWidget("buttongridcontainer");
    mysvc->AddChildBefore(mybuttongrid, lower);

}

void PeriodicApp::HandleAbuttongridEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;
    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uibuttongridcode" )
        {
            AppLog( "AButtonGrid support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_abuttongrid.cpp", widget->GetParent() );
        }
        else if (widget->GetId() ==  "uibuttongridlayout" )
        {
            AppLog( "AButtonGrid support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_abuttongrid.ui.txt", widget->GetParent() );
        }
        else
        {
           AppLog( "AButtonGrid action grid button : " + widget->GetId() + " was pressed ");
        }
    }
    if ( widget->GetId() ==  "AButtonGridDemo" )
    {
        AppLog( "AButtonGrid event : " + widget->GetId() + " got event= "+ EventReport(eventData[P_TYPE].GetInt()) );
    }
}

